### PR TITLE
refactor(ec2): simplify convert_ec2_metadata_network_config

### PR DESCRIPTION
**Pull request by aciba90**
See individual commits.

## Additional Context

https://warthogs.atlassian.net/browse/SC-1717

## Checklist

- [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/development/index.html)
- [x] I have updated or added any [unit tests](https://cloudinit.readthedocs.io/en/latest/development/testing.html) accordingly
- [x] I have updated or added any [documentation](https://cloudinit.readthedocs.io/en/latest/devlopment/contribute_docs.html) accordingly

## Merge type

- [ ] Squash merge using "Proposed Commit Message"
- [x] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

**Review by TheRealFalcon — APPROVED** [review]
LGTM!

File: cloudinit/sources/DataSourceEc2.py [hunk 1]
```diff
@@ -947,6 +948,85 @@ def _get_key_as_int_or(dikt, key, alt_value):
...
+            dev_config["routes"].append(
+                {
+                    "to": "0.0.0.0/0",
+                    "via": gateway,
+                    "table": table,
+                },
+            )
+
+    if is_ipv4:
+        subnet_prefix_routes = nic_metadata["subnet-ipv4-cidr-block"]
```
Comment (blackboxsw) [inline cloudinit/sources/DataSourceEc2.py:996]: Should this just be consolidated into the above is_ipv4 scope?
```suggestion
            )
        subnet_prefix_routes = nic_metadata["subnet-ipv4-cidr-block"]
```

or Maybe just this diff?
```diff
--- a/cloudinit/sources/DataSourceEc2.py
+++ b/cloudinit/sources/DataSourceEc2.py
@@ -971,6 +971,7 @@ def _configure_policy_routing(
     if not dev_config.get("routes"):
         dev_config["routes"] = []
     if is_ipv4:
+        subnet_prefix_routes = nic_metadata["subnet-ipv4-cidr-block"]
         try:
             lease = distro.dhcp_client.dhcp_discovery(nic_name, distro=distro)
             gateway = lease["routers"]
@@ -991,9 +992,6 @@ def _configure_policy_routing(
                     "table": table,
                 },
             )
-
-    if is_ipv4:
-        subnet_prefix_routes = nic_metadata["subnet-ipv4-cidr-block"]
     else:
         subnet_prefix_routes = nic_metadata["subnet-ipv6-cidr-blocks"]
     subnet_prefix_routes = (
```

File: cloudinit/sources/DataSourceEc2.py [hunk 2]
```diff
@@ -947,6 +948,85 @@ def _get_key_as_int_or(dikt, key, alt_value):
...
+        )
+
+    if not dev_config.get("routing-policy"):
+        dev_config["routing-policy"] = []
+    # Packets coming from any IP associated with the current NIC
+    # will be routed using `table` routing table
+    if is_ipv4:
+        ips = nic_metadata["local-ipv4s"]
+    else:
+        ips = nic_metadata["ipv6s"]
```
Comment (blackboxsw) [inline cloudinit/sources/DataSourceEc2.py:1019]: Can't this conditional also be pulled into that first `is is_ipv4 too?
```diff
--- a/cloudinit/sources/DataSourceEc2.py
+++ b/cloudinit/sources/DataSourceEc2.py
@@ -971,6 +971,8 @@ def _configure_policy_routing(
     if not dev_config.get("routes"):
         dev_config["routes"] = []
     if is_ipv4:
+        subnet_prefix_routes = nic_metadata["subnet-ipv4-cidr-block"]
+        ips = nic_metadata["local-ipv4s"]
         try:
             lease = distro.dhcp_client.dhcp_discovery(nic_name, distro=distro)
             gateway = lease["routers"]
@@ -991,11 +993,10 @@ def _configure_policy_routing(
                     "table": table,
                 },
             )
-
-    if is_ipv4:
-        subnet_prefix_routes = nic_metadata["subnet-ipv4-cidr-block"]
     else:
         subnet_prefix_routes = nic_metadata["subnet-ipv6-cidr-blocks"]
+        ips = nic_metadata["ipv6s"]
+
     subnet_prefix_routes = (
         [subnet_prefix_routes]
         if isinstance(subnet_prefix_routes, str)
@@ -1013,10 +1014,6 @@ def _configure_policy_routing(
         dev_config["routing-policy"] = []
     # Packets coming from any IP associated with the current NIC
     # will be routed using `table` routing table
-    if is_ipv4:
-        ips = nic_metadata["local-ipv4s"]
-    else:
-        ips = nic_metadata["ipv6s"]
     ips = [ips] if isinstance(ips, str) else ips
     for ip in ips:
         dev_config["routing-policy"].append(
```

Comment (aciba90) [reply cloudinit/sources/DataSourceEc2.py]: Applied, plus additional reference to GH issue.

**Comment by aciba90** [comment]
Thanks, @TheRealFalcon and @blackboxsw, for reviews and comments. I have addressed them.

**Review by blackboxsw — APPROVED** [review]
Thanks for the minor fixes. I've confirmed across upgrade path from daily PPA that we can now properly see ipv4/ipv6 route policies with this changeset on multi-nic dual-stack instances.
```diff
--- /etc/netplan/50-cloud-init.yaml	2024-03-19 19:12:34.256000129 +0000
+++ 50-cloud-init.yaml.PR5071	2024-03-19 19:00:12.680600947 +0000
@@ -18,9 +18,23 @@
             dhcp4: true
             dhcp4-overrides: &id002
                 route-metric: 200
+                use-routes: true
             dhcp6: true
             dhcp6-overrides: *id002
             match:
                 macaddress: 02:45:35:1b:63:4f
+            routes:
+            -   table: 101
+                to: 0.0.0.0/0
+                via: 192.168.0.1
+            -   table: 101
+                to: 192.168.0.0/20
+            -   table: 101
+                to: 2600:1f16:8cf:d900:0:0:0:0/64
+            routing-policy:
+            -   from: 192.168.4.35
+                table: 101
+            -   from: 2600:1f16:8cf:d900:8983:30de:9268:4a8f
+                table: 101
             set-name: ens6
     version: 2
```

**Pull request merged**